File: solutions/98 - Validate Binary Search Tree/dfs.cpp
#include <iostream>
#include <string>
#include <sstream>
#include <vector>
#include <queue>
#include <climits>
using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    bool isValidBST(TreeNode* root) {
        // Time Complexity: O(N) because we need to go through each node
        // Space Complexity: O(H) where H is the height of the tree for recursion stack
        return check(root, LONG_MIN, LONG_MAX);
    }
    bool check(TreeNode* root, long mini, long maxi)
    {
        if (root == nullptr)
        {
            return true;
        }
        if (root->val <= mini || root->val >= maxi)
        {
            return false;
        }
        return check(root->left, mini, root->val) && check(root->right, root->val, maxi);
    } 
};

int main() {
    cout << "Enter BST values: " << endl;

    string line;
    getline(cin, line);

    stringstream ss(line);
    vector<string> tokens;
    string tok;
    while (ss >> tok) tokens.push_back(tok);

    TreeNode* root = nullptr;

    if (!tokens.empty() && tokens[0] != "null") {
        root = new TreeNode(stoi(tokens[0]));
        queue<TreeNode*> q;
        q.push(root);

        size_t i = 1;
        while (!q.empty() && i < tokens.size()) {
            TreeNode* cur = q.front();
            q.pop();

            if (i < tokens.size() && tokens[i] != "null") {
                cur->left = new TreeNode(stoi(tokens[i]));
                q.push(cur->left);
            }
            i++;

            if (i < tokens.size() && tokens[i] != "null") {
                cur->right = new TreeNode(stoi(tokens[i]));
                q.push(cur->right);
            }
            i++;
        }
    }

    Solution sol;
    cout << "Is the tree a valid BST?: " << (sol.isValidBST(root) ? "true" : "false") << endl;

    return 0;
}

File: solutions/98 - Validate Binary Search Tree/DFS.java
import java.util.*;
// Definition for a binary tree node.
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    public boolean isValidBST(TreeNode root) {
        // Time Complexity: O(N) because we visit each node in the BST.
        // Space Complexity: O(H) for the recursion stack. O(N) worst case for a skewed tree.
        return dfs(root, Integer.MIN_VALUE, Integer.MAX_VALUE);
    }

    private boolean dfs(TreeNode root, int lower, int higher) {
        if (root == null) {
            return true;
        }
        // Check curr node with global bounds.
        if (root.val <= lower || root.val >= higher) {
            return false;
        }
        return dfs(root.left, lower, root.val) && dfs(root.right, root.val, higher);
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.println("Enter BST values: ");
        String line = sc.nextLine().trim();

        TreeNode root = null;

        if (!line.isEmpty()) {
            String[] tokens = line.split("\\s+");

            if (!tokens[0].equalsIgnoreCase("null")) {
                root = new TreeNode(Integer.parseInt(tokens[0]));
                Queue<TreeNode> q = new ArrayDeque<>();
                q.add(root);

                int i = 1;
                while (!q.isEmpty() && i < tokens.length) {
                    TreeNode cur = q.poll();

                    if (i < tokens.length && !tokens[i].equalsIgnoreCase("null")) {
                        cur.left = new TreeNode(Integer.parseInt(tokens[i]));
                        q.add(cur.left);
                    }
                    i++;

                    if (i < tokens.length && !tokens[i].equalsIgnoreCase("null")) {
                        cur.right = new TreeNode(Integer.parseInt(tokens[i]));
                        q.add(cur.right);
                    }
                    i++;
                }
            }
        }
        Solution sol = new Solution();
        System.out.println("Is the tree a valid BST?: " + sol.isValidBST(root));

        sc.close();
    }
}